### ISSUE TITLE
Interpreter behaves non-monotonically

### ISSUE THREAD

**Issue by ngeiswei** [trueagi-io/hyperon-experimental#461]
## What is the problem?

It's not clear if it is a bug or a feature, but adding a function definition may remove some branches of the non-determinism.

## How to reproduce the problem?

Run the following metta code

```clojure
!(let (A $x) ($f $x) (Result $f $x))
```
it outputs
```
[(Result A $x)]
```

Then add the following definition at the top

```clojure
(= (f $x) (A B))
```
now it outputs
```
[(Result f B)]
```

The result makes sense, however what makes less sense is that the previous result `(Result A $x)` has disappeared.

**Comment by vsbogd** [comment]
I would say in fact it is a duplicate of https://github.com/trueagi-io/hyperon-experimental/issues/242
@ngeiswei please re-open if you disagree.

**Comment by ngeiswei** [comment]
I don't think it is exactly a duplicate of #242.  When `variable_operation` is enabled it remains that the reduction of `($f $x)` is masking the unification of `($f $x)` with `(A $x)`, while I think it should create two branches instead.  But since `variable_operation` is disabled most of the time, maybe it is somewhat of a non-issue.

**Comment by ngeiswei** [comment]
I'm reopening is case this requires further discussion.  Feel free to close if you think that is a non-issue.

**Comment by ngeiswei** [comment]
I've created a test case that is independent of issue #242.

```clojure
;; (= (f $x) (A B))
!(let ($f A) (f $x) (Result $f $x))
```

outputs `[(Result f A)]`.  If `(= (f $x) (A B))` is enabled then it outputs `[]`.

But considering it again I'm even more confused whether it is a bug or a feature.

**Comment by vsbogd** [comment]
Reading your code and understanding that `let` reduces the second argument I see that code works as expected. When `f` definition is provided it calculates `f` and tries to match results. Otherwise it matches two expression without reduction.

Which behaviour do you expect in case when first line is uncommented? Do you think we could return both alternatives in this case?

**Comment by ngeiswei** [comment]
> Which behaviour do you expect in case when first line is uncommented? Do you think we could return both alternatives in this case?

I must admit that I am confused and I do not know what should be the correct behavior.  Maybe @Necr0x0Der has an opinion.

**Comment by vsbogd** [comment]
To elaborate it further. `let` has type `(-> Atom %Undefined% Atom Atom)` (or `(-> Atom %Undefined% Atom %Undefined%)` in Rust interpreter). It means the second argument is reduced before unification. If you want to prevent this you can use `quote`:
```
(= (f $x) (A B))
!(let (quote ($f A)) (quote (f $x)) (Result $f $x))
```
This code will always return `[(Result f A)]`

**Comment by Necr0x0Der** [comment]
Taking into account the current type signature of `let`, the behavior is indeed consistent, although may not be intuitive at the syntactic level. In contrast to `let`, `(unify (f $x) (f $x) $x False)` works, although with presence of `(= (f $x) (A B))` the result got reduced to `$x#5` instead of just `$x`.
The fact that `let` reduces its second argument is very convenient in many cases, and `let` is actually widely used to reduce arguments of functions, which accept `Atom`, when we want this. It would be a good topic for a tutorial, but I'm not sure that we will not redesign `Atom` type, or separate types and meta-types.
In any case, there is definitely no bug in this example. The confusion may come from dynamic typing and more than one tool for doing stuff. Right now, `quote` is an ordinary function, which utilizes `Atom` type. One option would be to not have `Atom` type at all (or do not treat it specially in terms of non-reduction of arguments), but to have `quote` not as an ordinary function, but as a built-in symbol, which tells the interpreter that the expression should not be reduced. In this case, you should have to always quote such expressions. It wouldn't provide additional capabilities beyond what we have we `quote` already, I believe, and it would make expression processing inconvenient. Special `Atom` type treatment was introduced exactly because we are working with expressions themselves a lot, and wrapping each expression in `quote` (possible in a nested way) would be really inconvenient.
The original issue was because of evaluation of `($f $x)`, which is kind of changed/fixed. But I'm not sure if non-monotonic behavior is avoidable in general.

**Comment by vsbogd** [comment]
May be I don't understand correctly what does "monotonic" mean here, but we expect MeTTa programs will change their behaviour if we are changing context. But may be more correct way of interpreting this example it to return both alternatives: `[(Result f A)]` and `[]`? I remember that similar question was already raised by @ngeiswei in https://github.com/trueagi-io/hyperon-experimental/issues/385#issuecomment-1665389199

**Comment by ngeiswei** [comment]
I mean monotonic as in https://en.wikipedia.org/wiki/Monotonicity_of_entailment

In general we wouldn't expect an AGI to be stuck with a monotonic logic, so I guess the question is: under which assumption should the rewriting system behave monotonically?  One could for instance assume that if `add-atom` or `remove-atom` are forbidden then such rewriting system should behave monotonically.  I'm not saying it should be the case, frankly I don't know, I'm just wondering.

**Comment by vsbogd** [comment]
> I suppose it is a different issue, which clearly relates to reduction control.

Yes, issue is different, but what I wanted to point out we can change the semantics of `(f $x)` in `let` to return both reduced and non-reduced version. I am not sure whether it will made interpreter more monotonic though.
